py/desisim/templates.py

```diff
@@ -1901,6 +1916,10 @@ def make_templates(self, nmodel=100, zrange=(0.5, 4.0), rmagrange=(20.0, 22.5),
                     flux[kk, :] = np.dot(self.eigenflux.T, PCA_rand[:, kk]).flatten()
                     if redshift[ii] > 2.39:
                          flux[kk, :pix912] *= np.exp(-phys_dist.value / mfp[ii])
+                    if bal: 
+                        if bal.isbal(balprob): 
+                            baltemplate, balindex[kk] = bal.getbaltemplate(zwave[:,ii], redshift[ii], balwave, baltemplates)
+                            flux[kk, :] *= baltemplate
```

Comment:
Do the BAL templates need to be computed one-by-one, in the loop, or can they be pre-computed in a way that's vectorized?  If all you need is the redshift and a probability that a given QSO has a BAL then I *think* you should be able to pre-generate these spectra, which should speed up this step measurably.

Reply:
The complete templates are now read in from the input file, which should speed this up.